### PR TITLE
[Dragon Hunt] Move Cura custom setting functions to a separate file

### PR DESCRIPTION
**Pull request by LipuFei**
@nallath Do some damage to the `ExtruderManager`.

### CODE REVIEW

**Comment by nallath** [comment]
Would it also be possible to add some unit tests to the functions you moved?

**Comment by LipuFei** [comment]
@nallath I'll look into it.

**Comment by LipuFei** [comment]
Pretty hard to make good tests right now. I'll skip that for now.

**Comment by LipuFei** [comment]
@nallath This one is ready for merge. I'll create new PRs for followups. Less big bang PRs to review.

File: cura/Settings/CustomSettingFunctions.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+from UM.Settings.PropertyEvaluationContext import PropertyEvaluationContext
+from UM.Settings.SettingFunction import SettingFunction
+
+if TYPE_CHECKING:
+    from cura.CuraApplication import CuraApplication
+    from cura.Settings.CuraContainerStack import CuraContainerStack
+
+
+#
+# This class contains all Cura-related custom setting functions. Some functions requires information such as the
```
Comment (nallath) [inline cura/Settings/CustomSettingFunctions.py]: Setting Function should be together (Since you are refering to UM.Settings.SettingFunction)

Comment (nallath) [reply cura/Settings/CustomSettingFunctions.py]: That's what a SettingFunction is. At least, that's what the documentation of the class says; "Encapsulates Python code that provides a simple value calculation function."

Comment (LipuFei) [reply cura/Settings/CustomSettingFunctions.py]: That's not correct. If a setting's property is a function, that field will be evaluated using `SettingFunction`. So, you can say that `SettingFunction` represents the value of the a setting property which is a Python code that needs to be evaluated. It does not represent any single (custom) function in a formula.

I'll update the documentation for this. As for `CustomSettingFunctions`, maybe `CustomFormulaFunctions` can be a more accurate name?

Comment (LipuFei) [reply cura/Settings/CustomSettingFunctions.py]: `SettingFunction` doc updated in https://github.com/Ultimaker/Uranium/commit/be7934eff93b50466c4dca8b3c165ed097f32372

**Comment by Appesteijn** [comment]
CURA-5781

**Comment by nallath** [comment]
Hmm. It's causing some issues with me though; 
2018-10-03 13:39:00,888 - DEBUG - [MainThread] UM.Logger.logException [84]: Exception: An exception occurred in inherit function =extruderValue(adhesion_extruder_nr, 'adhesion_type'): '0'
2018-10-03 13:39:00,892 - DEBUG - [MainThread] UM.Logger.logException [88]: Traceback (most recent call last):
2018-10-03 13:39:00,899 - DEBUG - [MainThread] UM.Logger.logException [88]:   File "/home/jaime/Development/Uranium/UM/Settings/SettingFunction.py", line 95, in __call__
2018-10-03 13:39:00,904 - DEBUG - [MainThread] UM.Logger.logException [88]:     return eval(self._compiled, g, locals)
2018-10-03 13:39:00,907 - DEBUG - [MainThread] UM.Logger.logException [88]:   File "<UM.Settings.SettingFunction (0x7fa88a791e10) =extruderValue(adhesion_extruder_nr, 'adhesion_type') >", line 1, in <module>
2018-10-03 13:39:00,912 - DEBUG - [MainThread] UM.Logger.logException [88]:   File "/home/jaime/Development/Cura/cura/Settings/CuraFormulaFunctions.py", line 41, in getValueInExtruder
2018-10-03 13:39:00,918 - DEBUG - [MainThread] UM.Logger.logException [88]:     extruder_stack = global_stack.extruders[str(extruder_position)]
2018-10-03 13:39:00,921 - DEBUG - [MainThread] UM.Logger.logException [88]: KeyError: '0'
2018-10-03 13:39:00,927 - INFO - [MainThread] cura.Settings.GlobalStack.addExtruder [88]: Extruder[ultimaker2_extruder_0 #2] added to [Ultimaker 2] at position [0]

**Comment by LipuFei** [comment]
Fixed that error in 7d7de32

**Pull request merged**